core/src/main/java/jenkins/fingerprints/FileFingerprintStorage.java

```diff
@@ -210,6 +217,61 @@ public boolean isReady() {
         return new File(Jenkins.get().getRootDir(),"fingerprints").exists();
     }
 
+    /**
+     * Perform Fingerprint cleanup.
+     */
+    public void iterateAndCleanupFingerprints(TaskListener taskListener) {
+        Object fingerprintStorage = FingerprintStorage.get();
+        if (!(fingerprintStorage instanceof FileFingerprintStorage)) {
+            logger.fine("External fingerprint storage is configured. Skipping execution");
+            return;
+        }
+
+        int numFiles = 0;
+
+        File root = new File(getRootDir(), FINGERPRINTS_DIR_NAME);
+        File[] files1 = root.listFiles(f -> f.isDirectory() && f.getName().length()==2);
+        if(files1!=null) {
+            for (File file1 : files1) {
+                File[] files2 = file1.listFiles(f -> f.isDirectory() && f.getName().length()==2);
+                for(File file2 : files2) {
+                    File[] files3 = file2.listFiles(f -> f.isFile() && FINGERPRINT_FILE_PATTERN.matcher(f.getName()).matches());
+                    for(File file3 : files3) {
+                        if(cleanFingerprint(file3, taskListener))
+                            numFiles++;
+                    }
+                    deleteIfEmpty(file2);
+                }
+                deleteIfEmpty(file1);
+            }
+        }
+
+        taskListener.getLogger().println("Cleaned up "+numFiles+" records");
+    }
+
+    private boolean cleanFingerprint(File fingerprintFile, TaskListener listener) {
+        try {
+            Fingerprint fp = loadFingerprint(fingerprintFile);
+            if (fp == null || (!fp.isAlive() && fp.getFacetBlockingDeletion() == null) ) {
+                listener.getLogger().println("deleting obsolete " + fingerprintFile);
+                fingerprintFile.delete();
+                return true;
+            } else {
+                if (!fp.isAlive()) {
+                    FingerprintFacet deletionBlockerFacet = fp.getFacetBlockingDeletion();
+                    listener.getLogger().println(deletionBlockerFacet.getClass().getName() + " created on " + new Date(deletionBlockerFacet.getTimestamp()) + " blocked deletion of " + fingerprintFile);
```

Comment:
```suggestion
                    listener.getLogger().println(String.format("%s created on %s blocked deletion of %s", deletionBlockerFacet.getClass().getName(), new Date(deletionBlockerFacet.getTimestamp()), fingerprintFile));
```

Reply:
This is moved code, so will make this change in separate PR.